test/e2e/persistent_volumes.go

```diff
@@ -47,16 +47,158 @@ func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
+func deletePVCandValidatePV(c *client.Client, ns string, pvc *api.PersistentVolumeClaim, pv *api.PersistentVolume) {
+
+	By("Deleting PersistentVolumeClaim to trigger PV Recycling")
+
+	framework.Logf("Deleting PersistentVolumeClaim %v to trigger PV Recycling", pvc.Name)
+	err := c.PersistentVolumeClaims(ns).Delete(pvc.Name)
+	if err != nil {
+		framework.Failf("Delete PersistentVolumeClaim %v failed: %v", pvc.Name, err)
+	}
+
+	// Wait for the PV's phase to return to Available
+	framework.Logf("Waiting for recycling process to complete.")
+	err = framework.WaitForPersistentVolumePhase(api.VolumeAvailable, c, pv.Name, 3*time.Second, 300*time.Second)
+	if err != nil {
+		framework.Failf("Recycling failed: %v", err)
+	}
+
+	// Examine the pv.ClaimRef and UID. Expect nil values.
+	pv, err = c.PersistentVolumes().Get(pv.Name)
+	if err != nil {
+		framework.Failf("Cannot re-get PersistentVolume %v:", pv.Name, err)
+	}
+	if pv.Spec.ClaimRef != nil && len(pv.Spec.ClaimRef.UID) > 0 {
+		crJSON, _ := json.Marshal(pv.Spec.ClaimRef)
+		framework.Failf("Expected PV %v's ClaimRef to be nil, or the claimRef's UID to be blank. Instead claimRef is: %v", pv.Name, string(crJSON))
+	}
+}
+
+// create the PV resource. Fails test on error.
+func createPV(c *client.Client, pv *api.PersistentVolume) (*api.PersistentVolume) {
+
+	pv, err := c.PersistentVolumes().Create(pv)
+	if err != nil {
+		framework.Failf("Create PersistentVolume %v failed: %v", pv.Name, err)
+		return nil
+	}
+
+	return pv
+}
+
+// create the PVC resource. Fails test on error.
+func createPVC(c *client.Client, ns string, pvc *api.PersistentVolumeClaim) (*api.PersistentVolumeClaim) {
+
+	pvc, err := c.PersistentVolumeClaims(ns).Create(pvc)
+	if err != nil {
+		framework.Failf("Create PersistentVolumeClaim %v failed: %v", pvc.Name, err)
+		return nil
+	}
+
+	return pvc
+}
+
+// Create a PV and PVC based on the passed in nfs-server ip and namespace. 
+// There are 4 combinations, 3 of which are supported here:
+//   1) prebind and create pvc first
+//   2) no prebind and create pvc first
+//   3) no prebind and create pv first
+// The case of prebinding and creating the pv first is not possible due to using a
+// generated name in the pvc, and thus not knowing the claim's name until after it
+// has been created.
+func createPVandPVC(c *client.Client, serverIP, ns string, pvFirst, preBind bool) (*api.PersistentVolume, *api.PersistentVolumeClaim) {
+
+	var bindTo *api.PersistentVolumeClaim
+
+	pvc := makePersistentVolumeClaim(ns) // pvc.Name not known yet
+
+	bindTo = nil
+	if preBind { // implies pvc *must* be created before the pv
+		pvFirst = false
+		bindTo = pvc
+	}
+	pv := makePersistentVolume(serverIP, bindTo)
+
+	if pvFirst {
+		By("Creating the PV followed by the PVC")
+		pv = createPV(c, pv)
+		pvc = createPVC(c, ns, pvc)
+	} else {
+		By("Creating the PVC followed by the PV")
+		pvc = createPVC(c, ns, pvc)
+		// need to fill-in claimRef with pvc.Name 
+		pv.Spec.ClaimRef.Name = pvc.Name
```

Comment:
@jsafrane pv.ClaimRef.name is not filled in (by makePersistentVolume) when the pv is created because the associated claim does not know its own name until the claim is instantiated. I believe this is due to using a generated named.